docs/tutorial/workarounds.md

```diff
@@ -0,0 +1,204 @@
+
+@fhe.compiler({"numbers": "encrypted", "threshold": "encrypted"})
+def filtering(numbers, threshold):
+    is_greater = numbers > threshold
+
+    shifted_numbers = numbers * 2  # open space for a single bit at the end
+    combined_numbers_and_is_greater = shifted_numbers + is_greater  # put is_greater to that bit
+
+    def extract(combination):
+        is_greater = (combination % 2) == 1  # extract is_greater back from packing
+        if_true = combination // 2  # if is greater is true, we unpack the number and use it
+        if_false = 0  # otherwise we set the element to zero
+        return np.where(is_greater, if_true, if_false)  # and apply the operation
+
+    return fhe.univariate(extract)(combined_numbers_and_is_greater)
+
+inputset = [(np.random.randint(0, 16, size=5), np.random.randint(0, 16)) for _ in range(50)]
+circuit = filtering.compile(inputset)
+
+numbers = np.random.randint(0, 16, size=5)
+threshold = np.random.randint(0, 16)
+assert np.array_equal(circuit.encrypt_run_decrypt(numbers, threshold), list(map(lambda x: x if x > threshold else 0, numbers)))
+
+```
+
+## Matrix Row/Col means
+
+In this example of Matrix operation, we are introducing a key concept when using `Concrete`:
+trying to maximize the parallelization. Here instead of sequentially sum all values to create a
+mean value, we split the values in sub-groups, and do the mean of the sub-groups means:
+
+```python
+import numpy as np
+from concrete import fhe
+
+def smallest_prime_divisor(n):
+    if n % 2 == 0:
+        return 2
+
+    for i in range(3, int(np.sqrt(n)) + 1):
+        if n % i == 0:
+            return i
+
+    return n
+
+def mean_of_vector(x):
+    assert x.size != 0
+    if x.size == 1:
+        return x[0]
+
+    group_size = smallest_prime_divisor(x.size)
+    if x.size == group_size:
+        return np.round(np.sum(x) / x.size).astype(np.int64)
+
+    groups = []
+    for i in range(x.size // group_size):
+        start = i * group_size
+        end = start + group_size
+        groups.append(x[start:end])
+
+    mean_of_groups = []
+    for group in groups:
+        mean_of_groups.append(np.round(np.sum(group) / group_size).astype(np.int64))
+
+    return mean_of_vector(fhe.array(mean_of_groups))
+
+@fhe.compiler(({"x": "encrypted"}))
+def mean_of_matrix(x):
+    return mean_of_vector(x.flatten())
+
+@fhe.compiler(({"x": "encrypted"}))
+def mean_of_rows_of_matrix(x):
+    means = []
+    for i in range(x.shape[0]):
+        means.append(mean_of_vector(x[i]))
+    return fhe.array(means)
+
+@fhe.compiler(({"x": "encrypted"}))
+def mean_of_columns_of_matrix(x):
+    means = []
+    for i in range(x.shape[1]):
+        means.append(mean_of_vector(x[:, i]))
+    return fhe.array(means)
+
+
+inputset = [np.random.randint(0, 16, size=(5,5)) for _ in range(50)]
+circuit = mean_of_matrix.compile(inputset)
+
+matrix = np.random.randint(0, 16, size=(5, 5))
+assert  circuit.encrypt_run_decrypt(matrix) == round(matrix.mean())
```

Comment:
should we have a test for mean_of_rows_of_matrix(?

Reply:
and for mean_of_columns_of_matrix

Reply:
added